app/views/automated_tests/student_interface.html.erb

```diff
@@ -4,15 +4,15 @@
 <%= javascript_include_tag 'result', nonce: true %>
 <%= javascript_include_tag 'Results/main', nonce: true %>
 
-<% if @assignment.enable_student_tests && !@grouping.nil? %>
+<% if policy(@student).run_tests? %>
   <p>
     <strong><%= Assignment.human_attribute_name(:token_start_date) %>:</strong>
     <%= l(@assignment.token_start_date) %>
   </p>
 
   <p>
     <strong><%= Assignment.human_attribute_name(:token_end_date) %>:</strong>
-    <%= l(@assignment.token_end_date) %>
+    <%= @assignment.token_end_date.present? ? l(@assignment.token_end_date) : l(@assignment.due_date) %>
```

Comment:
My previous comment on this line was about the use of `@assignment.due_date`. This doesn't correctly handle section-specific due dates or when students receive extensions.

You should instead use `@grouping.collection_date`. (This is seen by starting with the `run_tests?` policy, then going into the `before_due_date?` check for `grouping`, then the `Grouping#past_collection_date?` method.)